src/main/scala/ch/epfl/bluebrain/nexus/kg/resources/attachment/AttachmentStore.scala

```diff
@@ -0,0 +1,170 @@
+import cats.{Applicative, Monad}
+import ch.epfl.bluebrain.nexus.kg.config.AttachmentConfig
+import ch.epfl.bluebrain.nexus.kg.resources.Rejection.Unexpected
+import ch.epfl.bluebrain.nexus.kg.resources.attachment.Attachment._
+import ch.epfl.bluebrain.nexus.kg.resources.attachment.AttachmentStore.LocationResolver.Location
+import ch.epfl.bluebrain.nexus.kg.resources.attachment.AttachmentStore._
+import ch.epfl.bluebrain.nexus.kg.resources.{Rejection, ResId}
+import ch.epfl.bluebrain.nexus.rdf.Iri
+import ch.epfl.bluebrain.nexus.rdf.Iri.{AbsoluteIri, RelativeIri}
+
+import scala.concurrent.Future
+import scala.util.Try
+
+abstract class AttachmentStore[F[_]: Monad, In, Out](implicit loc: LocationResolver[F], stream: Stream[F, In, Out]) {
+
+  /**
+    * Stores the provided stream source delegating to ''locator'' for choosing the location
+    * and to ''fileStream'' for storing the source on the selected location.
+    *
+    * @param id       the id of the resource
+    * @param att              the attachment to be stored
+    * @param source           the source
+    * @return [[BinaryAttributes]] wrapped in the abstract ''F[_]''  type if successful,
+    *         or a [[ch.epfl.bluebrain.nexus.kg.resources.Rejection]] wrapped within ''F[_]'' otherwise
+    */
+  def save(id: ResId, att: BinaryDescription, source: In): EitherT[F, Rejection, BinaryAttributes] =
+    loc(id, att.uuid).flatMap(location => stream.toSink(location, source, att)).map(att.process)
+
+  /**
+    * Fetches the binary associated to the provided ''attachment'' delegating to ''locator'' for choosing the location
+    * where to retrieve it and to ''fileStream'' for retrieving it.
+    *
+    * @param attachment the attachment metadata
+    */
+  def fetch(attachment: BinaryAttributes): Either[Rejection, Out] =
+    stream.toSource(attachment.fileUri.resolve(loc.base))
+}
+
+object AttachmentStore {
+
+  type AkkaIn  = Source[ByteString, Any]
+  type AkkaOut = Source[ByteString, Future[IOResult]]
+
+  trait Stream[F[_], In, Out] {
+
+    /**
+      * Attempts to create a Out form a URI.
+      * This should be used to transmit the content referred by the URI through the Out type in streaming fashion.
+      *
+      * @param uri the AbsoluteIri from where to retrieve the content
+      * @return the typeclass Out
+      */
+    def toSource(uri: AbsoluteIri): Either[Rejection, Out]
+
+    /**
+      * Attempts to store and create metadata information which will be used by [[ch.epfl.bluebrain.nexus.kg.resources.State]]
+      * from an incoming source of type In (which is typically a stream).
+      *
+      * @param loc    the location of the attachment
+      * @param source the source
+      * @param meta   the source metadata
+      */
+    def toSink(loc: Location, source: In, meta: BinaryDescription): EitherT[F, Rejection, StoredSummary]
+  }
+
+  object Stream {
+    def akka(config: AttachmentConfig)(implicit as: ActorSystem): Stream[Future, AkkaIn, AkkaOut] =
```

Comment:
We should also provide a Stream[Task, AkkaIn, AkkaOut] that wraps the akka one, something like:
```
def task(config: AttachmentConfig)(implicit as: ActorSystem): Stream[Task, AkkaIn, AkkaOut] =
  new Stream[Task, AkkaIn, AkkaOut] {
    val underlying = akka(config)
    def toSource(uri: AbsoluteIri): Either[Rejection, AkkaOut] =
      underlying.toSource(uri)
    def toSink(...): EitherT[Task, Rejection, StoredSummary] = {
      EitherT(Task.deferFuture(underlying.toSink(...).value))
    }
  }
```
such that we can use it directly in the routes using the same `F`.

Reply:
We can do this separately, though.